### ISSUE TITLE
[Meta] Drupal 11 support

### ISSUE THREAD

**Issue by stephen-cox** [localgovdrupal/localgov#764]
Update: Closing in favour of drupal.org issue at https://www.drupal.org/project/localgov/issues/3566368 for remaining tasks

----
Creating this as a meta issue to track Drupal 11 support.

As discussed at Merge Tuesday, we need to get an idea of how big a job fully supporting Drupal 11 will be. Hopefully, the code should largely support D11 now, so this is likely largely come down to checking contrib support.

## Current state of Drupal 11 support:

Dependencies without Drupal 11 support:

* ~https://www.drupal.org/project/autocomplete_deluxe~ added in [2.10](https://www.drupal.org/project/autocomplete_deluxe/releases/2.1.0)
* ~https://www.drupal.org/project/fontawesome~ added in [3.0.0](https://www.drupal.org/project/fontawesome/releases/3.0.0)
* ~https://www.drupal.org/project/geo_entity~ added in [1.0.2](https://www.drupal.org/project/geo_entity/releases/1.0.2)
* ~https://www.drupal.org/project/image_widget_crop~ currently only in [dev](https://www.drupal.org/project/image_widget_crop/releases/8.x-2.x-dev)
* ~https://www.drupal.org/project/office_hours~ added in [1.19](https://www.drupal.org/project/office_hours/releases/8.x-1.19)
* ~https://www.drupal.org/project/role_delegation~ added in [8.x-1.3](https://www.drupal.org/project/role_delegation/releases/8.x-1.3)
* ~https://www.drupal.org/project/search_api_location~ added in [8.x-1.0-alpha4](https://www.drupal.org/project/search_api_location/releases/8.x-1.0-alpha4)
* ~https://www.drupal.org/project/social_post~ added in [3.0.2](https://www.drupal.org/project/social_post/releases/3.0.2)
* https://www.drupal.org/project/social_post_twitter
* ~https://www.drupal.org/project/tablefield~ added in [3.0.0-beta1](https://www.drupal.org/project/tablefield/releases/3.0.0-beta1)
* ~https://www.drupal.org/project/views_field_view~ added in [8.x-1.0-beta6](https://www.drupal.org/project/views_field_view/releases/8.x-1.0-beta6)
* ~https://www.drupal.org/project/webform~ added in [6.3.0-alpha2](https://www.drupal.org/project/webform/releases/6.3.0-alpha2)
* ~https://www.drupal.org/project/group (dev dependency on a few modules e.g. alert banner)~ - added in [3.3.0](https://www.drupal.org/project/group/releases/3.3.0)
* ~https://www.drupal.org/project/facets_form (dependency of directories)~ - added in [1.2.0](https://www.drupal.org/project/facets_form/releases/1.2.0)

Also see https://docs.google.com/spreadsheets/d/1QVDrKtLI1KuAeMfZRQpn0O7nCRX7w7fo0StFHXP58g8/edit?gid=0#gid=0

Issues:

* https://github.com/localgovdrupal/localgov/issues/766
* https://github.com/localgovdrupal/localgov_alert_banner/issues/334
* https://github.com/localgovdrupal/localgov_blogs/issues/39
* https://github.com/localgovdrupal/localgov_core/issues/246
* https://github.com/localgovdrupal/localgov_directories/issues/394
* https://github.com/localgovdrupal/localgov_elections/issues/130
* https://github.com/localgovdrupal/localgov_events/issues/158
* https://github.com/localgovdrupal/localgov_forms/issues/90
* https://github.com/localgovdrupal/localgov_guides/issues/177
* https://github.com/localgovdrupal/localgov_news/issues/122
* https://github.com/localgovdrupal/localgov_openreferral/issues/56
* https://github.com/localgovdrupal/localgov_page_components/issues/59
* https://github.com/localgovdrupal/localgov_paragraphs/issues/193
* https://github.com/localgovdrupal/localgov_search/issues/76
* https://github.com/localgovdrupal/localgov_search_solr/issues/20
* https://github.com/localgovdrupal/localgov_services/issues/280
* https://github.com/localgovdrupal/localgov_step_by_step/issues/114
* https://github.com/localgovdrupal/localgov_subsites/issues/158
* https://github.com/localgovdrupal/localgov_workflows/issues/114

PRs:

* https://github.com/localgovdrupal/localgov_claro/pull/23
* https://github.com/localgovdrupal/localgov_demo/pull/139
* https://github.com/localgovdrupal/localgov_events/pull/157
* https://github.com/localgovdrupal/localgov_geo/pull/130
* https://github.com/localgovdrupal/localgov_menu_link_group/pull/42
* https://github.com/localgovdrupal/localgov_netcall/pull/6
* https://github.com/localgovdrupal/localgov_page/pull/56
* https://github.com/localgovdrupal/localgov_sa11y/pull/37
* https://github.com/localgovdrupal/localgov_search/pull/77
* https://github.com/localgovdrupal/localgov_search_solr/pull/21
* https://github.com/localgovdrupal/localgov_subsites_extras/pull/35
* https://github.com/localgovdrupal/localgov_topics/pull/31
* https://github.com/localgovdrupal/localgov_core/pull/262
* https://github.com/localgovdrupal/localgov_directories/pull/408
* https://github.com/localgovdrupal/localgov_openreferral/pull/59

**Comment by stephen-cox** [comment]
@millnut I believe you have started this. Can you add an update on how far you've got with this?

I'll find some time to look at this, but I don't want to duplicate work that has been done so far.

**Comment by millnut** [comment]
Yep sure I've got all this compiled and should be able to post something for discussion.

As you mentioned @stephen-cox the biggest problem is where contrib modules have two versions, one for D10 and one for D11

**Comment by millnut** [comment]
@stephen-cox I've added a 4.x branch for the project which set drupal core to 11 and updates the workflow for just 4.x requirements https://github.com/localgovdrupal/localgov_project/tree/4.x I'm expecting this to fail to start with

**Comment by millnut** [comment]
@stephen-cox here is the file I've been looking at with notes on contrib module changes
https://docs.google.com/spreadsheets/d/1QVDrKtLI1KuAeMfZRQpn0O7nCRX7w7fo0StFHXP58g8/edit?usp=sharing be good if we can upload to Google Docs so we can collab on it

**Comment by stephen-cox** [comment]
Thanks @millnut for the spreadsheet.

Looking at this it looks like the main D11 upgrade is blocked by the following contrib modules:

* [image_widget_crop](https://www.drupal.org/project/image_widget_crop/issues/3451140) - require by core.
* [role_delegation](https://www.drupal.org/project/role_delegation/issues/3434267) - required by core.
* [fontawesome](https://www.drupal.org/project/fontawesome/issues/3465765) - require by paragraphs.
* [office_hours](https://www.drupal.org/project/office_hours/issues/3433670) - required by paragraphs.
* [tablefield](https://www.drupal.org/project/tablefield/issues/3451341) - required by paragraphs.
* [search_api_location](https://www.drupal.org/project/search_api_location/issues/3455298) - required by directories.

There are other contrib dependencies blocking upgrades to some of the other modules, but these are included by default.

We either need to wait for D11 support to be added to these modules or agree a workaround for them.

**Comment by millnut** [comment]
Thanks for looking through @stephen-cox I might look at contributing some patches on drupal.org for those and see if they get picked up

**Comment by markconroy** [comment]
Acquia have a portal here to show the status of all modules.

https://dev.acquia.com/drupal11/deprecation_status/projects

Here's the modules we are waiting for: https://dev.acquia.com/drupal11/deprecation_status/projects?names=image_widget_crop%2C%20role_delegation%2C%20fontawesome%2C%20office_hours%2C%20tablefield%2C%20search_api_location

**Comment by stephen-cox** [comment]
Looks like we're now waiting for Drupal 11 support in:

* [fontawesome](https://www.drupal.org/project/fontawesome/issues/3465765)
* [office_hours](https://www.drupal.org/project/office_hours/issues/3433670)
* [tablefield](https://www.drupal.org/project/tablefield/issues/3451341)

All required by paragraphs.

**Comment by millnut** [comment]
I have updated the first post and the spreadsheet with the latest contrib module compatibility

**Comment by millnut** [comment]
Existing installs using composer aliases for D11 support is making progress and composer installs work (see https://github.com/localgovdrupal/localgov_shared_workflows/pull/19#issuecomment-2561885845) just need to update our modules and wait for `fontawesome` and `image_widget_crop` to tag releases as these are the main blockers. For testing we are blocked waiting for a new release of `caxy/php-htmldiff` but that is specifically for PHP 8.4 testing (have patched for now).

Composer aliases so far;

```bash
"localgovdrupal/localgov"
"localgovdrupal/localgov_core"
"localgovdrupal/localgov_demo"
"localgovdrupal/localgov_directories"
"localgovdrupal/localgov_events"
"localgovdrupal/localgov_geo"
"localgovdrupal/localgov_guides"
"localgovdrupal/localgov_news"
"localgovdrupal/localgov_page_components"
"localgovdrupal/localgov_paragraphs"
"localgovdrupal/localgov_services"
"localgovdrupal/localgov_step_by_step"
"localgovdrupal/localgov_subsites"
"localgovdrupal/localgov_workflows"
```

The remaining one is uninstalled by default but part of core is `localgov_openreferral` which needs quite a bit of work. Who maintains this one now?

We can start reviewing these WIP D11 PRs, merging, and creating releases on low-impact ones.

New/Clean installs and installs via Drush there is still some work to do on module core version compatibility as well as the contrib modules above.

**Comment by millnut** [comment]
`fontawesome` should have a release soon it went to the project update working group https://www.drupal.org/project/puwg/issues/3498021 for a chase on D11 release date

**Comment by millnut** [comment]
I've contacted a maintainer of `image_widget_crop` on rough ETA for a D11 compatible tagged version

**Comment by millnut** [comment]
@ekes @stephen-cox directories is blocked for D11 by `drupal/facets_form` which doesn't have a D11 release will add to the list

**Comment by andybroomfield** [comment]
Facets form has a D11 fix here https://www.drupal.org/project/facets_form/issues/3455218 (labbled as needs review)
As its a dev only dependency (to run tests with the facets form plugin) maybe a case to use Drupal lienient and patch? to get tests to pass?

**Comment by ekes** [comment]
I was going to ask why facets_form. But yes it's only a dev dependency. It's also RTBC by one of the maintainers now, but not worth delaying on that.

**Comment by millnut** [comment]
@finnlewis @ekes @stephen-cox @andybroomfield This is my recent update/check for D11;

### Core project (order to be merged in)

1. directories
2. events
3. demo
4. profile
5. project

### Modules outside of core project (still needs work)

- blogs
- elections
- forms
- openreferral (supports D11 but has a lot of code style issues)
- there may be others I'm not aware of...

### Contrib for core project
- image_widget_crop still no 2.x release for D11 but there is a 3.0.0 release, might need to update to that and re-test

I've excluded microsites for now as that probably needs to be tracked separately

**Comment by finnlewis** [comment]
- Directories released: https://github.com/localgovdrupal/localgov_directories/releases/tag/3.3.0
- Events released: https://github.com/localgovdrupal/localgov_events/releases/tag/3.0.4
- Demo released: https://github.com/localgovdrupal/localgov_demo/releases/tag/3.0.2

Still need to look at image_widget_crop bump.

Then @stephen-cox will look at a beta or RC release of localgov and localgov_project to be ready for testing!

:)

**Comment by willguv** [comment]
Alpha release out there, requires councils to test:

- @finnlewis to test at Greenwich
- @stephen-cox to test on demo.localgovdrupal
- Core team think about how much testing = beta release
- - @finnlewis to use new tech mailing list, and suggest a timeframe
- @TimDCR think about how else we can encourage councils to test

**Comment by millnut** [comment]
Just added some D11 issues we'll need to resolve before Beta release - https://github.com/issues?q=is%3Aissue+user%3Alocalgovdrupal+is%3Aopen+label%3A%22drupal+11%22+

**Comment by willguv** [comment]
Agree a route to Beta

@finnlewis to

- test upgrade path
- demo the upgrade in a community session or at DrupalCamp?

@stephen-cox says there are no critical bugs so should be Beta

Finn/ Stephen to create MRs, propose it at Merge Tuesday/ on Slack

**Comment by willguv** [comment]
- [x] Agree a route to Beta and release
- [x] @finnlewis test upgrade path
- [x] @willguv encourage councils to try it
- [ ] Agree in Tech Governance how to review stable blockers and then do the work
- [ ] @stephen-cox check what's left on Elections, Open Referral. Ekes says OR is OK
- [ ] Check remaining issues https://github.com/issues?q=is%3Aissue+user%3Alocalgovdrupal+is%3Aopen+label%3A%22drupal+11%22+
- [ ] @finnlewis demo the upgrade in a community session (June) or at DrupalCamp (3 July)

**Comment by willguv** [comment]
See @stephen-cox's Elections social post module

https://github.com/localgovdrupal/localgov_elections/issues/159

**Comment by willguv** [comment]
- @finnlewis @stephen-cox what's the path to a stable release: tech gov group conversation

**Comment by msayoung** [comment]
Stephen : 
- agreed to spit out localgov_elections (social post module causing issues)  
- there are tests failing due to core problems in 11.2 , so planning to pin to 11.1 for the time being

**Comment by tonypaulbarker** [comment]
localgov_elections work is approved -> 

We have this remaining issue to look at https://www.drupal.org/project/drupal/issues/3529670

**Comment by tonypaulbarker** [comment]
Discussion ongoing here regarding the media module https://github.com/localgovdrupal/localgov_core/issues/307

**Comment by finnlewis** [comment]
This is the Drupal.org issue where we're exploring if this is a core bug or a problem with our config dependencies.
https://www.drupal.org/project/drupal/issues/3529670

**Comment by tonypaulbarker** [comment]
I am blocked from progress on 11.2 support https://github.com/localgovdrupal/localgov_core/issues/307  . I would like to:
- establish whether or not the key allows project installs to run through the UI / drush (we have mixed reports - if we do get mixed reports, what conditions could be different?)
- find out _why_ config validation fails for unit tests in 11.2 (what changed in core, can we pinpoint the core changes between 11.1 and 11.2) - we suspect something in or related to https://www.drupal.org/project/drupal/issues/3416522 - 

I think that config dependencies are calculated before hook_modules_installed is called whether container_rebuild_required is true or not for test scenarios and I think that possibly container_rebuild_required doesn't have any effect at all for timing in tests, that possibly tests can't see the hook, that possibly validation became more strict. I'm sure that a heavyweight back end developer will have a better understanding of this than I do.

We _could_ make some blunt changes to move forward but understanding what's going on will help inform next steps.

**Comment by willguv** [comment]
@tonypaulbarker to set up PR with workaround

**Comment by tonypaulbarker** [comment]
Slight change of tact as from last round of testing we might be closer than we thought to solving it.

**Comment by millnut** [comment]
I believe I've fixed the remaining issues preventing installs with Drupal 11.2

**Comment by millnut** [comment]
@tonypaulbarker around the PHPUnit test failures, I've been able to run a full test with Drupal 11.2 now the install errors are resolved. This now highlights these errors https://github.com/localgovdrupal/localgov_project/actions/runs/17008752320/job/48222015066.

What is interesting is for each failure, if we set the profile to `localgov` like below, the test then passes.

```php
  /**
   * {@inheritdoc}
   */
  protected $profile = 'localgov';
```

Maybe we do this for the failing tests, and then look into recipes for managing the case of the `config/optional` in `localgov_media` and adopt recipes and move away from the profile

**Comment by willguv** [comment]
Planning 4 Sep - aim for a release candidate this month

**Comment by willguv** [comment]
@tonypaulbarker @finnlewis to update on progress here

**Comment by willguv** [comment]
@tonypaulbarker to note the remaining couple of issues so we can focus on them

**Comment by stephen-cox** [comment]
@tonypaulbarker @millnut @finnlewis It looks to me from Lee's last test run that Drupal 11.2 tests are still failing on the config in the media module: https://github.com/localgovdrupal/localgov_project/actions/runs/17650801726/job/50161163704

One option to fix this that occurs to me is we can create a module with the config we need just for testing. Tests can then add this as a dependency and the tests will pass.

There's a branch with this here: https://github.com/localgovdrupal/localgov_core/tree/feature/media-test-config

I've created a PR in Events to check this: https://github.com/localgovdrupal/localgov_events/pull/203

This is a bit of a hack, but might be a quick way to get the tests passing so we can push on with a Drupal 11 release while we work on the underlying problem.

Thoughts?

**Comment by millnut** [comment]
I agree with you @stephen-cox this seems the most sensible approach with defining this test module as a testing dependency only on affected modules. We can then unpick this into recipes in future.

**Comment by tonypaulbarker** [comment]
I had a brief look at unpicking the conflict and I didn't spot the problem. I have been working on the configuration so resolving the underlying issue may win the race if we can't get the tests to pass with the old code.

**Comment by stephen-cox** [comment]
I have created a PR that adds a media test config module to LGD Core: https://github.com/localgovdrupal/localgov_core/pull/320

If we're happy to merge and release this then I can go through all the modules with failing tests on media config and add this module as a dependency. This should get the tests passing so we can move on with getting a stable D11 release out.

**Comment by tonypaulbarker** [comment]
Thanks @stephen-cox that sounds like a plan!

**Comment by willguv** [comment]
@tonypaulbarker please update

**Comment by willguv** [comment]
@tonypaulbarker to catch up on progress and update

**Comment by tonypaulbarker** [comment]
Closing in favour of drupal.org issue at https://www.drupal.org/project/localgov/issues/3566368 for remaining tasks
